src/backend/cuda/kernel/moments.hpp

```diff
@@ -0,0 +1,94 @@
+/*******************************************************
+ * Copyright (c) 2016, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <af/defines.h>
+#include <backend.hpp>
+#include <Param.hpp>
+#include <dispatch.hpp>
+#include <math.hpp>
+#include <err_cuda.hpp>
+#include <debug_cuda.hpp>
+
+namespace cuda
+{
+namespace kernel
+{
+
+    // Kernel Launch Config Values
+    static const int THREADS = 128;
+
+    template<typename T>
+    __global__
+    void moments_kernel(CParam<float> out, CParam<T> in, af_moment_type moment,
+                  const dim_t blocksMatX, const bool pBatch)
+    {
+        const dim_t idw = blockIdx.y / in.dims[2];
+        const dim_t idz = blockIdx.y - idw * in.dims[2];
+
+        const dim_t idy = blockIdx.x;
+        dim_t idx = threadIdx.x;
+
+        __shared__ float blk_moment_sum[4];
+        if(threadIdx.x < 4) {
+            blk_moment_sum[threadIdx.x] = 0.f;
+        }
+        __syncthreads();
+
+        for(unsigned i=0; i<blocksMatX; ++i) {
+            dim_t mId = idy * in.strides[1] + idx;
+            if(pBatch) {
+                mId += idw * in.strides[3] + idz * in.strides[2];
+            }
+
+            if (idx >= in.dims[0] || idy >= in.dims[1] ||
+                idz >= in.dims[2] || idw >= in.dims[3] )
+                break;
```

Comment:
You can check for the `id{y,z,w}` conditions before entering the loop and only check for `idx` in here.